### ISSUE TITLE
Performance Reports

### ISSUE THREAD

**Issue by mokemokechicken** [mokemokechicken/reversi-alpha-zero#40]
Please share your reversi model achievement!
Battle record, configuration, repository url, comments and so on.

**Comment by mokemokechicken** [comment]
There are lots of information [here](https://github.com/mokemokechicken/reversi-alpha-zero/issues/26).

**Comment by gooooloo** [comment]
I have some model reports in `README` of this [repo](https://github.com/gooooloo/reversi-alpha-zero)

**Comment by richemslie** [comment]
[I have started a recent run](https://github.com/ggplib/ggp-zero/blob/dev/doc/reversi_record.md), This is using [ggp-zero](https://github.com/ggplib/ggp-zero) (reversi-alpha-zero implementation was inspiration!).  ggp-zero is a generic implementation of a 'zero' method, and can train many different games.  ie By zero I mean start with a random network and train via self play using (PUCT or variant) MCTS.  However, at this point the implementation (and goals) are very divergent from AlphaZero (I also drew inspiration from 'Thinking Fast and Slow').  For this run, I am running with multiple policies and multiple value heads, with no turn flipping of the network and no symmetry/rotation of the network.

A previous run achieved approximately ntest level 7, however there were no records.  Will keep up to date this time!

**Comment by apollo-time** [comment]
@gooooloo what's mean ntest:6 result (6/1/3) for step-418500?

**Comment by gooooloo** [comment]
@apollo-time in my report, 6/1/3 means 6 wins, 1 draw, 3 lose. By the way, you can find the game saves in https://github.com/gooooloo/reversi-alpha-zero-models/tree/master/ggf )

**Comment by gooooloo** [comment]
@apollo-time 

> How about policy and value loss? 

policy loss : 0.4
value loss : 0.4-0.55. Unstable.

**Comment by apollo-time** [comment]
@gooooloo Um...But you use game history, isn't it?

**Comment by gooooloo** [comment]
@apollo-time 

> Um...But you use game history, isn't it?

Yes. I am actually guessing that, the resnet blocks number in the model should be reduced depends on the historical board number: Less historical boards, shallower the model should be. There is bigger chance to overfit if network's input space is large. I sometimes observe strange, bad move from my model, right after several strong moves. That I cannot understand. Overfitting could be a explain, yet I cannot make sure.

**Comment by richemslie** [comment]
I finished up my latest run, ending up somewhere between ntest 5-10, depending on the phase of the moon.  Not too shabby. 
The policy loss was about 2.0 and value loss is 0.14 - I found the above numbers very interesting.
@gooooloo - congrats on creating such a strong model.  Can you try ntest level 7 again with your latest model, I would love to see it getting beat? :)

**Comment by gooooloo** [comment]
@richemslie congrats on getting the result. ntest5 is already strong, I feel. 

> Can you try ntest level 7 again with your latest model

I have detailed evaluation metric [here](https://github.com/gooooloo/reversi-alpha-zero/blob/master/records.md). The model beats ntest6 6/1/3 is step-418500+400sim, it loses to ntest7... But it beats ntest9 with 7/0/3. It seems to me that winning ntest:x doesn't imply winning ntext:(x-1) or ntest:(x-2)...

**Comment by apollo-time** [comment]
@gooooloo I see your player gametreee and see you subtract virtual loss after all simulations(800).
Why do you not use parallel_search_num and is it ok?

**Comment by gooooloo** [comment]
@apollo-time 

>  subtract virtual loss after all simulations(800)

Do you mean I should subtract virtual loss every simulation, not until all sims finished? That's what I do inside backup(). My function naming is not accurate though, will refine them later...

> Why do you not use parallel_search_num and is it ok?

I just change the implementation to single thread way 3 days ago. I think the efficiency is 
not changed and codes are simpler to read, and I can also use it for time control (@mokemokechicken  pointed it out the coroutine way can also do time control, as in this [issue](https://github.com/mokemokechicken/reversi-alpha-zero/issues/42) ). 

Yet my model was most trained from this [version](https://github.com/gooooloo/reversi-alpha-zero/blob/9471b944650b2cad9f769812eaefa03744a0665e/src/reversi_zero/agent/player.py) of GameTree implementation. You check it either.

**Comment by apollo-time** [comment]
@gooooloo Oh, I see you subtract virtual loss inside backup().
I am checking now why my model can't beat ntest and windows reversi though low policy and value loss.
I'll check your previous version. thanks.

**Comment by gooooloo** [comment]
Report: my step-473800 model with 30 minutes per game time control beats ntest lv7-10, draw lv11, beats lv12-13, lose to lv14( 2 draws, 2 loses). Game model can be found [here](https://github.com/gooooloo/reversi-alpha-zero-models/tree/master/model_473800-steps), evaluation game savings can be found [here](https://github.com/gooooloo/reversi-alpha-zero-models/tree/master/ggf/step-473800/min-30)

Why use 30 min? Actually I am targeting 5 min C++ implementation on a 8-core CPU, 1-GPU machine. But I don't have a C++ implementation yet, so I try current python implementation with 30 min. I believe they will be similar in simulation_number_per_move, which is about 13000. And I believe 5 minutes of C++ implementation is reasonable for evaluating AI's length.

**Comment by mokemokechicken** [comment]
@gooooloo 
That's interesting.
In my self play, there was a time when the draw was very much(more than 50%).
There was also a time when black's winning percentage was more than 50%, and vice versa.

I think one of the advantage of AlphaZero is to change models always.
However, one of the disadvantage is that it is easy to overfit the training data and to become weak.
So I feel that "training/self-play ratio" of #38 is very important.

**Comment by fuzzthink** [comment]
Wondering which is currently the strongest model. Is Challenge 1 still the strongest? Also, `sh ./download_model.sh 2` seems to save to `data/model/model_best_*`, but config.py (and in README) seems to expect it in `/data/model/next_generation/*`

**Comment by richemslie** [comment]
Hi - new [record](https://github.com/ggplib/ggp-zero/blob/dev/doc/reversi_record.md) for gzero (of a different kind - playing equal to ntest level 3 after *only 12* hours of training).  Discovered a pretty bad bug with PUCT constants this morning, which is the reason for a new run.  

Some points:
 * Used a tiny->small network (gen34 had 8 residual layers, 96 filters and 128 hidden before value head) - which is one reason could train so fast.
 * gen 35 was much weaker than 34, which is making me consider adding an evaluation stage before switching out network
 * I trained using 2 gpus (1080/1080ti) - and have 6k concurrent self-play games - I swap out the network mid game each new generation, and since gen35 was weaker than gen34 - this could potentially be harmful.
 * Only taking 6 samples per game.

**Comment by AranKomat** [comment]
@richemslie Your implementation using C++ looks very interesting. I'm curious how many self-play games per second or minute you can generate. By taking into account your small architecture, it may be possible to compare your speed with that of Python-only implementations.

**Comment by richemslie** [comment]
@AranKomat - it is hard to give exact numbers for comparison, but using a batch size of 1024 on 1080ti card, it can be 100% saturated with 2 c++ threads.   I am seeing 20480 model evaluations, and hence for ~60 moves at 800 evaluations per move, that is 2.3 seconds per game.  For the tiny (initial) network (8 res, 64 filters,64 hidden), it took 3 c++ threads to saturate and was about twice as fast (1.1 seconds).
The second card is throttled to prevent overheating it (adding a sleep into my optimised c++ code was painful!) - it roughly works out to 14 seconds per game.
Note that the reversi game is defined in a prolog like language, and then interpreted... so if it was a custom reversi implementation in c++ it would be much faster, and it wouldn't require so many threads to saturate the GPU.

**Comment by AranKomat** [comment]
@mokemokechicken achieved 22 seconds per game with 256 filters, a single 1080ti and half as many sims/move. I assume it would take 22s/game with two 1080ti's and 800 sims/move. Though arch FLOPS is (256/96)^2~7 times larger, GPU and TF scales weirdly. So, if I assume yours would take 2.3*3 s/game with the same architecture as moke's, the speedup of using C++ is probably 22/(2.3*3)~3.3 times? I'm looking forward to your updates!

**Comment by mokemokechicken** [comment]
@fuzzthink 

I am sorry to have confused you.

> Wondering which is currently the strongest model. Is Challenge 1 still the strongest? 

Now, "challenge 5 model" and "ch5 config" are strongest in my models.

> Also, sh ./download_model.sh 2 seems to save to data/model/model_best_*, but config.py (and in README) seems to expect it in /data/model/next_generation/*

Please remove( or rename) `data/model/next_generation/` directory if you want to use "BestModel" at `data/model/model_best_*`.

For example, 

```bash
rm -rf data/model/next_generation/
sh ./download_model.sh 5
# run as wxPython GUI
python src/reversi_zero/run.py play_gui --type ch5
```

If you want to use as a NBoard engine, please use `nboard_engine --type ch5` for the Command.
